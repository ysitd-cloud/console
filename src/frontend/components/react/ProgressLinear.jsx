import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

function ProgressLinearBar({
  active, indeterminate, bufferValue, children,
}) {
  const classNames = classnames('progress-linear__bar');
  const style = {};

  if (!active) {
    style.height = 0;
  }

  if (!indeterminate && bufferValue !== 100) {
    style.width = `${this.bufferValue}%`;
  }

  return <div className={classNames} style={style}>{children}</div>;
}

ProgressLinearBar.propTypes = {
  active: PropTypes.bool.isRequired,
  indeterminate: PropTypes.bool.isRequired,
  bufferValue: PropTypes.number.isRequired,
  children: PropTypes.node,
};

ProgressLinearBar.defaultProps = {
  children: null,
};

function ProgressLinearBackground({
  active, color, backgroundColor, backgroundOpacity, bufferValue,
}) {
  const classNames = classnames(backgroundColor || color, 'progress-linear__background');

  let opacity = backgroundOpacity;
  if (opacity === null) {
    opacity = backgroundColor ? 1 : 0.3;
  }

  const style = {
    opacity,
    height: active ? 'auto' : 0,
    width: `${bufferValue}%`,
  };

  return (
    <div className={classNames} style={style} />
  );
}

ProgressLinearBackground.propTypes = {
  active: PropTypes.bool.isRequired,
  color: PropTypes.string.isRequired,
  backgroundColor: PropTypes.string,
  backgroundOpacity: PropTypes.number,
  bufferValue: PropTypes.number.isRequired,
};

ProgressLinearBackground.defaultProps = {
  backgroundOpacity: null,
  backgroundColor: null,
};

function Indeterminate({ color }) {
  return (
    <div className="progress-linear__bar__indeterminate progress-linear__bar__indeterminate--active">
      <div className={classnames('progress-linear__bar__indeterminate', 'long', color)} />
      <div className={classnames('progress-linear__bar__indeterminate', 'short', color)} />
    </div>
  );
}

Indeterminate.propTypes = {
  color: PropTypes.string.isRequired,
};

function Determinate({ color, width }) {
  const style = {
    width: `${width}%`,
  };

  return (
    <div className={classnames('progress-linear__bar__determinate', color)} style={style} />
  );
}

Determinate.propTypes = {
  color: PropTypes.string.isRequired,
  width: PropTypes.number.isRequired,
};

export default function ProgressLinear({
  active,
  bufferValue,
  height,
  color,
  indeterminate,
  backgroundColor,
  backgroundOpacity,
  value,
}) {
  const style = {
    height: `${height}px`,
  };
  const classNames = 'progress-linear__bar';

  // eslint-disable-next-line no-mixed-operators
  const effectiveWidth = bufferValue ? (value * 100 / bufferValue) : 0;

  return (
    <div className={classNames} style={style}>
      <ProgressLinearBackground
        active={active}
        bufferValue={bufferValue}
        color={color}
        backgroundColor={backgroundColor}
        backgroundOpacity={backgroundOpacity}
      />
      <ProgressLinearBar
        active={active}
        indeterminate={indeterminate}
        bufferValue={bufferValue}
      >
        {indeterminate && <Indeterminate color={color} />}
        {!indeterminate && <Determinate color={color} width={effectiveWidth} />}
      </ProgressLinearBar>
    </div>
  );
}

ProgressLinear.propTypes = {
  height: PropTypes.number,
  color: PropTypes.string,
  backgroundColor: PropTypes.string,
  backgroundOpacity: PropTypes.number,
  bufferValue: PropTypes.number,
  active: PropTypes.bool,
  indeterminate: PropTypes.bool,
  value: PropTypes.number,
};

ProgressLinear.defaultProps = {
  height: 7,
  color: 'primary',
  backgroundColor: null,
  backgroundOpacity: null,
  bufferValue: 100,
  active: true,
  indeterminate: false,
  value: 0,
};
