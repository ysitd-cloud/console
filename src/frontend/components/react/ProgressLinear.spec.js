import React from 'react';
import renderer from 'react-test-renderer';
import ProgressLinear from './ProgressLinear';

/* globals describe, expect, it */

describe('React ProgressLinear Component', () => {
  it('match snapshot', () => {
    const component = renderer.create(<ProgressLinear indeterminate />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
