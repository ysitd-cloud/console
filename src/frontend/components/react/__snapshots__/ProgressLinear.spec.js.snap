// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React ProgressLinear Component match snapshot 1`] = `
<div
  className="progress-linear__bar"
  style={
    Object {
      "height": "7px",
    }
  }
>
  <div
    className="primary progress-linear__background"
    style={
      Object {
        "height": "auto",
        "opacity": 0.3,
        "width": "100%",
      }
    }
  />
  <div
    className="progress-linear__bar"
    style={Object {}}
  >
    <div
      className="progress-linear__bar__indeterminate progress-linear__bar__indeterminate--active"
    >
      <div
        className="progress-linear__bar__indeterminate long primary"
      />
      <div
        className="progress-linear__bar__indeterminate short primary"
      />
    </div>
  </div>
</div>
`;
